File: src/main/java/org/modelingvalue/nelumbo/strings/Length.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.strings;

import java.io.Serial;
import java.math.BigInteger;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.InferContext;
import org.modelingvalue.nelumbo.InferResult;
import org.modelingvalue.nelumbo.Predicate;
import org.modelingvalue.nelumbo.patterns.Functor;

public final class Length extends Predicate {
    @Serial
    private static final long serialVersionUID = 4405805306602130025L;

    public Length(Functor functor, List<AstElement> elements, Object[] args) {
        super(functor, elements, args[0], args[1]);
    }

    private Length(Object[] array, Length declaration) {
        super(array, declaration);
    }

    @Override
    protected Length struct(Object[] array, Predicate declaration) {
        return new Length(array, (Length) declaration);
    }

    @Override
    protected InferResult infer(int nrOfUnbound, InferContext context) {
        if (nrOfUnbound > 1) {
            return unknown();
        }

        java.lang.String string = getVal(0, 0);
        BigInteger length = getVal(1, 0);
        if (string != null) {
            BigInteger actual = BigInteger.valueOf(string.length());
            if (length != null) {
                boolean eq = length.equals(actual);
                return eq ? factCC() : falsehoodCC();
            } else {
                return set(1, org.modelingvalue.nelumbo.integers.Integer.of(actual)).factCI();
            }
        }

        return unknown();
    }

}

File: web-ts/src/strings/Length.ts
/**
 * Length - String length predicate.
 * @JAVA_REF org.modelingvalue.nelumbo.strings.Length
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Predicate } from '../logic/Predicate';
import type { Functor } from '../patterns/Functor';
import type { InferContext } from '../InferContext';
import type { InferResult } from '../InferResult';
import { NInteger } from '../integers/NInteger';
import { NString } from './NString';

export class Length extends Predicate {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, args[0], args[1]);
  }

  protected static fromDataLength(data: unknown[], declaration?: Predicate): Length {
    const node = Object.create(Length.prototype) as Length;
    (node as unknown as { _data: unknown[] })._data = data;
    (node as unknown as { _declaration: Predicate | undefined })._declaration = declaration ?? node;
    (node as any)._binding = null;
    (node as any)._hashCodeCached = false;
    (node as any)._hashCode = 0;
    (node as any)._nrOfUnbound = -1;
    return node;
  }

  protected override struct(data: unknown[], declaration?: Predicate): Predicate {
    return Length.fromDataLength(data, declaration ?? this.declaration());
  }

  // @JAVA_REF org.modelingvalue.nelumbo.strings.Length#infer(int, InferContext)
  protected override inferInternal(_nrOfUnbound: number, _context: InferContext): InferResult {
    const nrOfUnbound = this.nrOfUnbound();
    if (nrOfUnbound > 1) {
      return this.unresolvable();
    }

    const string = this.getStringVal(0);
    const length = this.getIntegerVal(1);

    if (string !== null) {
      const actual = BigInt(string.length);
      if (length !== null) {
        const eq = length === actual;
        return eq ? this.factCC() : this.falsehoodCC();
      } else {
        return (this.set(1, NInteger.of(actual)) as Length).factCI();
      }
    }

    return this.unknown();
  }

  private getStringVal(index: number): string | null {
    const val = this.get(index);
    if (val instanceof NString) {
      return val.value();
    }
    return null;
  }

  private getIntegerVal(index: number): bigint | null {
    const val = this.get(index);
    if (val instanceof NInteger) {
      return val.value();
    }
    return null;
  }
}
